ENH: stats: Updates to `variation`.

WarrenWeckesser opened the pull request:
* Move variation to its own file.
* Added the parameter `keepdims` to `stats.variation`.
* Updated the examples in the docstring.
* Added validation of the `axis` argument, so an invalid value
  raises `np.AxisError`.
* For `nan_policy="omit"`, use `np.nanstd` and `np.nanmean`
  instead of converting the input to a masked array.  This
  prevents the surprising return of a masked array from the
  function.
* Ensure that no spurious numpy warnings are generated for cases
  that we want to handle cleanly.
* Handle the case a.size == 0 as a special case to avoid spurious
  warnings.
* Moved `variation` unit tests to their own class, in their own
  file, replaced some expected values for existing tests with the
  exact value, and added tests for the new features.
* Add some utility functions to scipy/_lib/_util.py

Closes gh-592.

WarrenWeckesser commented:
Is the test that scans the code for uses of warning filters still necessary?  We now require Python 3.6 or later.

Also, the message about using `numpy.testing.suppress_warnings` says it is for use in tests only:

```
E   AssertionError: warning ignore filter should not be used, instead, use
E   numpy.testing.suppress_warnings (in tests only);
```
That imples we shouldn't use `numpy.testing.suppress_warnings` outside of the unit tests.

WarrenWeckesser commented:
Note: I included "Closes gh-592" in the commit message, but this PR doesn't implement the option for including the bias correction factor `1 + 4/n` that is mentioned in that issue.   If this PR is merged, I'll create a new issue specifically for that enhancement request, to be implemented later.

WarrenWeckesser commented:
It appears that, except for the standard floating point warnings, one cannot filter numpy warnings in a way that is thread-safe.  I pushed an update that instead of filtering warnings, jumps through the appropriate hoops to avoid generating the warnings in the first place.

WarrenWeckesser commented:
Still WIP--I haven't touched the `mstats` version.

WarrenWeckesser commented:
After a bit more work, I've marked this as ready for review.  I haven't updated `mstats.variation`; that can be done in a separate PR.

It probably looks like a lot of code to compute what is, in effect, `np.std(x) / np.mean(x)`.  But we want to handle edge cases and also `nan_policy='omit'` with a "nice" API (i.e. no spurious warnings), and the naive implementation doesn't give us that.

I put some potentially useful helper functions in `scipy/_lib/_util.py`.  As we add `nan_policy` to more functions, I suspect we'll develop more utility functions functions like these.  For now, `_lib._util` seems as good a place as any to put them.


The Travis CI failures are the known `linalg` failures of `test_heequb`.

WarrenWeckesser commented:
Rebased on master.  The one test failure on Travis CI looks unrelated.

WarrenWeckesser commented:
The test failure is https://github.com/scipy/scipy/issues/11281, which is not related to the PR.

rlucas7 commented on scipy/_lib/_util.py:
```diff
@@ -480,3 +482,98 @@ def rng_integers(gen, low, high=None, size=None, dtype='int64',
+    return res
+
+
+def _first_nonnan(a, axis):
```
Not super familiar with `_lib` setup, do we ***need*** to keep the function names "private" when the sub-package itself is already private?

WarrenWeckesser replied:
This is just "an overabundance of caution".  Suppose this function is imported and used in, say, `stats.py`.  Then someone could do `from scipy.stats.stats import _first_nonnan`.  Without the underscore prefix, it might then appear to be a public function.  Maybe that's a stretch, but that's why I refer to using the underscore as an overabundance of caution.

rlucas7 replied:
ok, thanks for clarifying, let's leave as is, better safe than sorry.

rlucas7 commented on scipy/stats/_variation.py:
```diff
@@ -0,0 +1,211 @@
+    """
+    Compute the coefficient of variation.
+
+    The coefficient of variation is the ratio of the standard deviation
+    to the mean.
```
is it worthwhile to write the eqn here to resolve any ambiguity on the reader/user?

WarrenWeckesser replied:
Sure, I'll add that in an update.

WarrenWeckesser replied:
Since this comment was made, the docstring had been updated with the formula (expressed with numpy function calls). I've updated the docstring in the new file with the same changes.

rlucas7 commented on scipy/stats/_variation.py:
```diff
@@ -0,0 +1,211 @@
+
+    std_a = a.std(axis, ddof=ddof, keepdims=True)
+
+    with np.errstate(divide='ignore', invalid='ignore'):
+        result = std_a / mean_a
```
this context manager would ignore the division by zero error. A user could (reasonably IMO) call this on data which has zero mean, not realize, and assume errors are propogated, investigate and get confused. Would it be better to explicitly state the behavior of division by 0 mean in the docstring (in a single sentence)?

WarrenWeckesser replied:
Yes, good idea.  I'll add a comment about what happens when the mean is 0, and in the special case where the input is all zeros (in which case the return value is `nan`).

WarrenWeckesser replied:
Done.

rlucas7 commented on scipy/stats/tests/test_variation.py:
```diff
@@ -0,0 +1,134 @@
+        assert_equal(y, np.nan)
+
+    @pytest.mark.parametrize('axis, expected',
+                             [(0, []), (1, [np.nan]*3), (None, np.nan)])
```
I'm not sure here so I'll ask, is it worthwhile to add a case w/just `inf`s here too?

WarrenWeckesser replied:
Do you mean the input is `inf`, or the result is `inf`?

rlucas7 replied:
> Do you mean the input is `inf`, or the result is `inf`?

I was thinking when the input is `inf`, I'd guess in that case the expected output is `nan` which is why I was asking. If there is also a case with `inf` outputs that makes sense in the context of the test sure but the specifics of that input are not apparent to me at this time.

WarrenWeckesser replied:
@rlucas7, I'm reviving this PR and catching up on some old comments.

An `inf` in the input will result in the output being `nan`.  The call to `numpy.std` will generate a warning in this case: `RuntimeWarning: invalid value encountered in subtract` that isn't very helpful.  I'll avoid that warning, and add a note that an `inf` in the input results in `nan`.

WarrenWeckesser replied:
Done.

rlucas7 commented on scipy/stats/tests/test_variation.py:
```diff
@@ -0,0 +1,134 @@
+from scipy.stats import variation
+
+
+class TestVariation:
```
From what I can tell the `variation` method in numpy is numerically stable, using 2 pass w/sample mean calculations. So it isn't an immediate concern but would be good to add a case where `std` has potential to overflow if unstable textbook calculation is used. 

Think that is worthwhile for maintenance ease or am I being excessive if we already have a stable variance calculation method?

WarrenWeckesser replied:
Hmmm... there are cases where `np.std()` will overflow and return `inf` even though the answer can be represented with 64 bit floating point.  E.g. ideally the following should return `2.5e200`:

```
In [27]: np.std([1e200, -2e200, 2e200, 5e200])                                  
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:195: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
Out[27]: inf
```
Because `variation` ultimately depends on `numpy.std`, it will also fail with this data:
```
In [61]: a = np.array([1e200, -2e200, 2e200, 5e200])

In [62]: variation(a)                                                           
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:195: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
Out[62]: inf
```
The correct answer is `2.5e200/1.5e200 = 1.66666....`.

A little algebra with the formula for the variation leads to this version (for 1-d data and no handling of all the various edge-cases or `nan_policy`):

```
In [63]: def v2(a, ddof=0): 
    ...:     a = np.asarray(a) 
    ...:     m = a.mean() 
    ...:     b = a/m 
    ...:     return np.std(b, ddof=ddof) 
    ...:  
```
This version *does* handle those large values:
```
In [64]: v2(a)
Out[64]: 1.6666666666666663
```
I'll investigate this some more, and see if it is worthwhile using that version as the basis for our implementation.

rlucas7 replied:
I think in this scenario you’d be maintaining precision with large stddev and large mean but would lose precision with the large mean small stddev cases because of the division by a large number. I haven’t tested this out but if my intuition is correct one thought is to branch the calculation logic based on some threshold for the mean (perhaps of the absolute values). Looking forward to see what you find in your experiments.

WarrenWeckesser replied:
This is a bit of a rabbit hole.  I think we're OK using `np.std(x)` for now, and leaving the alternative that I mentioned above for future investigation.

rlucas7 commented on scipy/stats/tests/test_variation.py:
```diff
@@ -0,0 +1,134 @@
+        with pytest.raises(ValueError, match='input contains nan'):
+            variation(x, nan_policy='raise')
+
+    def test_bad_nan_polciy(self):
```
```suggestion
    def test_bad_nan_polciy(self):
    def test_bad_nan_policy(self):
```

WarrenWeckesser replied:
Ooops.  I'll fix this in the next update.

WarrenWeckesser replied:
Done.

WarrenWeckesser replied:
Done.

rlucas7 commented on scipy/_lib/_util.py:
```diff
@@ -327,7 +328,7 @@ def getfullargspec_no_self(func):
         p.default for p in sig.parameters.values()
         if (p.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD and
-           p.default is not p.empty)
+            p.default is not p.empty)
```
these are (unrelated) pep8 cleanups, right?

WarrenWeckesser replied:
Yes.  Since there were just a couple issues found by `flake8`, I included the PEP-8 fixes along with my other changes.

rlucas7 replied:
ok, they look fine to me, just wanted to confirm my understand of why the changes were in the PR.

tylerjereddy commented:
@WarrenWeckesser @rlucas7 PR will be ready before we branch on Monday?

Not much activity here in quite some time, but comment from Lucas above seemed positive:

> found 1 typo and 1 open question on a test case (whether we want to include or not), looks like this PR is in good shape.

WarrenWeckesser commented:
@tylerjereddy, thanks for the ping.  I'll update this PR today.

WarrenWeckesser commented:
I don't know why the `Mypy` step of the **Linux Tests / Nightly CPython (3.10)** fails with the following error.  We use `normalize_axis_index` in many places.  
```
Run python -u runtests.py --mypy
  python -u runtests.py --mypy
  shell: /usr/bin/bash -e {0}
Building, see build.log...
    ... build in progress (0:01:00.076158 elapsed)
<snip>
    ... build in progress (0:21:01.517327 elapsed)
Build OK (0:21:49.575138 elapsed)
scipy/stats/_variation.py:3: error: Module "numpy.core.multiarray" has no attribute "normalize_axis_index"  [attr-defined]
Found 1 error in 1 file (checked 829 source files)
Error: Process completed with exit code 1.
```

WarrenWeckesser commented:
I "fixed" the mypy issue by adding the new file to the `mypy.ini` file configured to ignore errors (like we do with most of the other files).

I updated the PR with a merge of the master branch, and addressed @rlucas7's comments from last year with a few more changes.

tylerjereddy commented on scipy/stats/_variation.py:
```diff
@@ -0,0 +1,223 @@
+    return result
+
+
+def variation(a, axis=0, nan_policy='propagate', *, ddof=0, keepdims=False):
```
isn't this backwards incompatible with a positional use of `ddof` with the old function signature?

WarrenWeckesser replied:
Yes, good catch.  `ddof` was added to `variation` after the initial version of this pull request. I'll fix  that.

WarrenWeckesser replied:
Done.

tylerjereddy reviewed (commented):
Sounds like this is close, but I did add a question about backwards compatibility. Would be great if a `stats` regular confirmed they are "ok" with this, since it does modify some old tests it seems.

Also, not sure how important, but one of the code paths involving the new function parameter seems to be untested:

![image](https://user-images.githubusercontent.com/7903078/144731103-ea5fa278-2df3-4735-9a45-2df8355077dc.png)

WarrenWeckesser commented:
Thanks @tylerjereddy.  I pushed a change that addresses your comments.

>  it does modify some old tests it seems.

To help anyone reviewing, here are the details of the changes to the old tests.

There were just three old tests: `test_variation()`, `test_variation_propagate_nan()`, and `test_variation_ddof()`.

<details><summary>For convenience, here is a copy of the old tests.</summary>

```python
    def test_variation(self):
        # variation = samplestd / mean
        y = stats.variation(self.scalar_testcase)
        assert_approx_equal(y, 0.0)
        y = stats.variation(self.testcase)
        assert_approx_equal(y, 0.44721359549996, 10)

        x = np.arange(10.)
        x[9] = np.nan
        assert_equal(stats.variation(x), np.nan)
        assert_almost_equal(stats.variation(x, nan_policy='omit'),
                            0.6454972243679028)
        assert_raises(ValueError, stats.variation, x, nan_policy='raise')
        assert_raises(ValueError, stats.variation, x, nan_policy='foobar')

    def test_variation_propagate_nan(self):
        # Check that the shape of the result is the same for inputs
        # with and without nans, cf gh-5817
        a = np.arange(8).reshape(2, -1).astype(float)
        a[1, 0] = np.nan
        vv = stats.variation(a, axis=1, nan_policy="propagate")
        np.testing.assert_allclose(vv, [0.7453559924999299, np.nan], atol=1e-15)

    def test_variation_ddof(self):
        # test variation with delta degrees of freedom
        # regression test for gh-13341
        a = array([1, 2, 3, 4, 5])
        nan_a = array([1, 2, 3, np.nan, 4, 5, np.nan])
        y = stats.variation(a, ddof=1)
        nan_y = stats.variation(nan_a, nan_policy="omit", ddof=1)
        assert_approx_equal(y, 0.5270462766947299)
        np.testing.assert_equal(y, nan_y)
```
`self.scalar_testcase` is 4.0 and `self.testcase` is [1, 2, 3, 4].

</details>

All the conditions checked in the old tests are also checked in the new tests.

The old test `test_variation()` tests:
* a scalar input, which is now covered by `TestVariation.test_scalar()`
* a 1-d sequence, which is now covered by multiple tests, e.g. `TestVariation.test_ddof()` and `TestVariation.test_sign()`.
* a 1-d sequence containing `nan`, with different `nan_policy` arguments. The default case (i.e. `nan_policy='propagate'`) and `nan_policy='omit'` are now covered by `TestVariation.test_variation_nan()`.  The case `nan_policy='raise'` is now covered by `TestVariation.test_nan_policy_raise()`. The case `nan_policy='foobar'` is now covered by `TestVariation.test_bad_nan_policy`.

The old test `test_variation_propagate_nan()` tests `nan_policy='propagate'` with a 2-d array containing one `nan`.  This is now `TestVariation.test_propagate_nan()`.

The old test `test_variation_ddof()` tests that two arrays, one with no `nan`s and one with the same values along with a couple `nan`s produce the same result when the second array is processed with `nan_policy='omit'` and both use `ddof=1`.  This test is now `TestVariation.test_variation_ddof()`.

WarrenWeckesser commented:
Whenever this gets merged (whether by Monday or later for 1.9), please squash the commits, and use this as the single commit message:

```
ENH: stats: Updates to variation.

* Move variation to its own file.
* Added the parameter `keepdims` to `stats.variation`.
* Updated the examples in the docstring.
* Added validation of the `axis` argument, so an invalid value
  raises `np.AxisError`.
* For `nan_policy="omit"`, use `np.nanstd` and `np.nanmean`
  instead of converting the input to a masked array.  This
  prevents the surprising return of a masked array from the
  function.
* Ensure that no spurious numpy warnings are generated for cases
  that we want to handle cleanly.
* Handle the case a.size == 0 as a special case to avoid spurious
  warnings.
* Moved `variation` unit tests to their own class, in their own
  file, replaced some expected values for existing tests with the
  exact value, and added tests for the new features.
* Add some utility functions to scipy/_lib/_util.py
```
Or ping me, and I'll squash and force-push before the PR is committed.

tylerjereddy commented:
Thanks Warren et al.

I squash merged with the provided commit message on the following basis:

- the uncovered line I was worried about is now covered
- backward compatibility concern I had was fixed
- looks like the comments from Lucas, who regularly reviews our stats PRs, have all been addressed, if only by doc adjustments in some cases; Lucas had also mentioned that the PR was basically in pretty good shape already
- CI appears to be in pretty solid shape

h-vetinari commented:
PS. This closed the last remaining open three-digit issue 🥳
